### ISSUE TITLE
Font Optimization is not fully working

### ISSUE THREAD

**Issue by Harry-Yep** [vercel/next.js#35835]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000
    Binaries:
      Node: 17.8.0
      npm: 8.5.5
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 12.1.4
      react: 18.0.0
      react-dom: 18.0.0
```

### What browser are you using? (if relevant)

Chrome 100.0.4896.60

### How are you deploying your application? (if relevant)

Vercel, Heroku

### Describe the Bug

Font Optimization is not working fully.

This happened after upgrading to React@18. 

I have set up the project and using _document.tsx would like to add some of Google Fonts.

This is what it looks like in _document.tsx.
<img width="1463" alt="CleanShot 2022-04-02 at 17 07 48@2x" src="https://user-images.githubusercontent.com/66008528/161393640-0f66175f-e18b-4dae-99df-9f283f054025.png">

This is what it should look like after deployment.
<img width="1464" alt="CleanShot 2022-04-02 at 17 09 31@2x" src="https://user-images.githubusercontent.com/66008528/161393682-8b8cb5ed-d0e7-4f1e-aa1f-3597b33d1f5e.png">

However, this is what it looks currently like after deployment.
<img width="1365" alt="CleanShot 2022-04-02 at 17 10 30@2x" src="https://user-images.githubusercontent.com/66008528/161393706-adea89a5-eea8-4ca7-bfe4-2be68d84503f.png">

The `href` is turned to `data-href` after font optimization. But, `link` is not turned to `style`.

### Expected Behavior

Font Optimization should work.

### To Reproduce

Since it appears in individual projects, it is normal in most cases. So it's hard to reproduce. I tried, but I can't reproduce it by creating a new project.

**Comment by firatciftci** [comment]
I am also experiencing the same issue. This problem started after upgrading my project to React 18.0.0.

**Comment by irekrog** [comment]
Same issue, after upgraded to React 18 external fonts doesn't work

**Comment by balazsorban44** [comment]
Related: https://github.com/vercel/next.js/issues/24781#issuecomment-1088547525

**Comment by firatciftci** [comment]
@balazsorban44 My `_document.js` file is in the class component format and I am stil experiencing the issue.

**Comment by Harry-Yep** [comment]
Hi, @balazsorban44. I'm having the same situation as @firatciftci. So, it is may not be because `class` or `function` cause this issue.
Besides, as @irekrog said, when using `next dev`, font optimization works, however, it doesn't work for `next build`.

**Comment by irekrog** [comment]
Yep, and I've just noticed one thing. Fonts doesn't work when the page includes `getServerSideProps`. With static pages using `getStaticProps` or without any data fetching function fonts works fine.

**Comment by ronsj** [comment]
Experience the same issue in the latest version of `next`.

From a fresh install and copying exactly what is suggested in the [Font Optimization docs](https://nextjs.org/docs/basic-features/font-optimization). The `Inter` font does not display.

I ended up switching the `display` param in the Google Fonts URL from `optional` to `swap` just to get it to show the correct font.

**Comment by Cianekjr** [comment]
Temporary fix:
https://nextjs.org/docs/basic-features/font-optimization#disabling-optimization
Add: 
`optimizeFonts: false,`
to your next.config.js.

Then it won't change href to data-href.

**Comment by cleberj** [comment]
Seems like https://github.com/vercel/next.js/blob/4fb0beb1ee372d56128db16ec02ff88e333d4b2a/packages/next/server/render.tsx#L1310
we always have generateStaticHTML as false.
![image](https://user-images.githubusercontent.com/69492574/167339219-04005d3b-fa56-4286-ba0b-7e1c75c7442b.png)

Hence the processors never work for ssr:

![image](https://user-images.githubusercontent.com/69492574/167339343-2d589dfe-cfea-4a9e-8f3a-3076b84609dd.png)


There were some commits 3 months ago regarding this. Not sure if that is what make SSR stop working. I change the code manually to start as TRUE, and the web optimizations started working again.

**Comment by cleberj** [comment]
@shuding Do you know if this has sth to do with your refactoring? Can you help us ?

https://github.com/vercel/next.js/blob/41c0a66902899f7a085d4ec87c884ca4e5fff5c8/packages/next/build/webpack/loaders/next-middleware-ssr-loader/render.ts#L66

**Comment by huozhi** [comment]
Hi, the fix is landed in `next@12.1.7-canary.5`, could you try it to see if the issue is reolsved for you?

**Comment by Harry-Yep** [comment]
Hi @huozhi 

I can confirm that this issue has been resolved in `next@12.1.7-canary.5`.

Thanks
